core/lib/zksync_core/src/commitment_generator/mod.rs

```diff
@@ -0,0 +1,230 @@
+                    .await?;
+
+            let mut state_diffs = Vec::new();
+            for (key, value) in touched_slots {
+                let hashed_key = key.hashed_key();
+                let prev_value = previous_values[&hashed_key].unwrap_or_default();
+                if prev_value != value {
+                    let (initial_write_l1_batch_number, index) =
+                        l1_batches_for_initial_writes[&hashed_key];
+                    if initial_write_l1_batch_number == l1_batch_number {
+                        state_diffs.push(StateDiffRecord {
+                            address: *key.address(),
+                            key: h256_to_u256(*key.key()),
+                            derived_key: StorageKey::raw_hashed_key(key.address(), key.key()),
+                            enumeration_index: 0u64,
+                            initial_value: U256::default(),
+                            final_value: h256_to_u256(value),
+                        });
+                    } else {
+                        state_diffs.push(StateDiffRecord {
+                            address: *key.address(),
+                            key: h256_to_u256(*key.key()),
+                            derived_key: StorageKey::raw_hashed_key(key.address(), key.key()),
+                            enumeration_index: index,
+                            initial_value: h256_to_u256(prev_value),
+                            final_value: h256_to_u256(value),
+                        });
+                    }
+                }
+            }
+            state_diffs.sort_unstable_by_key(|rec| (rec.address, rec.key));
+
+            CommitmentInput::PostBoojum {
+                common,
+                system_logs: header.system_logs,
+                state_diffs,
+                aux_commitments,
+            }
+        };
+
+        Ok(input)
+    }
+
+    async fn step(
+        connection: &mut StorageProcessor<'_>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<()> {
+        let latency =
+            METRICS.generate_commitment_latency_stage[&CommitmentStage::PrepareInput].start();
+        let input = Self::prepare_input(connection, l1_batch_number).await?;
+        latency.observe();
+
+        let latency =
+            METRICS.generate_commitment_latency_stage[&CommitmentStage::Calculate].start();
+        let commitment = L1BatchCommitment::new(input);
+        let artifacts = commitment.artifacts();
+        latency.observe();
+
+        let latency =
+            METRICS.generate_commitment_latency_stage[&CommitmentStage::SaveResults].start();
+        connection
+            .blocks_dal()
+            .save_l1_batch_commitment_artifacts(l1_batch_number, &artifacts)
+            .await?;
+        latency.observe();
+
+        Ok(())
+    }
+
+    pub async fn run(self) -> anyhow::Result<()> {
+        let mut connection = self
```

Comment:
AFAIU, having a single long-living connection for the entire duration of the app lifetime is an anti-pattern; it's better to have shorter-lived connections, so that pooled connections can be utilized more efficiently.